utils/safeList.js

```diff
@@ -0,0 +1,286 @@
+    ETag: safeExtract,
+    Size: safeExtractInt,
+    StorageClass: safeExtract,
+    Owner: extractOwner,
+};
+
+const deleteMarkerKeys = {
+    Key: safeExtract,
+    VersionId: safeExtract,
+    IsLatest: safeExtractBool,
+    LastModified: safeExtractDate,
+    ETag: safeExtract,
+    Owner: extractOwner,
+    StorageClass: safeExtract,
+};
+
+const commonPrefixKeys = {
+    Prefix: safeExtract,
+};
+
+
+/**
+ * recoverListing - parse an xml listing response in a safe manner
+ * @param {object} resp - response from complete event of listObjectVersions
+ * @param {object|undefined} resp.httpResponse - http response from S3 (if any)
+ * @param {Error|undefined} resp.error - error from S3 (if any)
+ * @param {function} cb - callback
+ * @returns {undefined}
+ */
+function recoverListing(resp, cb) {
+    const parser = new xml2js.Parser();
+    parser.parseString(resp.httpResponse.body, (err, data) => {
+        if (err) {
+            // if there is an error parsing the xml, return the original error
+            return cb(resp.error);
+        }
+
+        if (!data.ListVersionsResult) {
+            // if no listing data is present, return the original error
+            return cb(resp.error);
+        }
+
+        const { value, error } = extractKeys(data.ListVersionsResult, listingKeys);
+        if (error) {
+            return cb(error);
+        }
+
+        const { value: versions, error: verErr } = safeExtractArray(
+            'Version',
+            data.ListVersionsResult,
+            versionKeys,
+        );
+        if (verErr) {
+            return cb(verErr);
+        }
+
+        const { value: deleteMarkers, error: delErr } = safeExtractArray(
+            'DeleteMarker',
+            data.ListVersionsResult,
+            deleteMarkerKeys,
+        );
+        if (delErr) {
+            return cb(delErr);
+        }
+
+        const { value: commonPrefixes, error: comErr } = safeExtractArray(
+            'CommonPrefix',
+            data.ListVersionsResult,
+            commonPrefixKeys,
+        );
+        if (comErr) {
+            return cb(comErr);
+        }
+
+        value.Versions = versions;
+        value.DeleteMarkers = deleteMarkers;
+        value.CommonPrefixes = commonPrefixes;
+        return cb(null, value);
+    });
+}
+
+/**
+ *  safeListObjectVersions - listObjectVersions with error recovery
+ *
+ * This function is a wrapper around listObjectVersions that will attempt to
+ * recover from certain format errors by parsing the XML response and returning the
+ * data in the same format as listObjectVersions.
+ *
+ * Values that fail to parse will be returned as the raw string value present in the xml.
+ *
+ * @param {AWS.S3} s3 - S3 client instance
+ * @param {object} params - listObjectVersions params
+ * @param {function} cb - callback
+ * @returns {undefined}
+ */
+function safeListObjectVersions(s3, params, cb) {
+    const req = s3.listObjectVersions(params);
+    req.on('complete', resp => {
```

Comment:
Don't we also have 'error' events to consider here?

Reply:
The "complete" event is fired when the request completes, regardless of whether there was an error.